File: graphs/L00.java
import java.util.ArrayList;
import java.util.LinkedList;

public class L00 {
    public static class Edge {
        int v = 0;
        int w = 0;

        Edge(int v, int w) {
            this.v = v;
            this.w = w;
        }
    }

    public static void addEdge(ArrayList<Edge>[] gp, int u, int v, int w) {
        gp[u].add(new Edge(v, w));
        gp[v].add(new Edge(u, w));
    }

    public static int findEdge(ArrayList<Edge>[] gp, int u, int v) {
        int i = 0;
        for (i = 0; i < gp[u].size(); i++) {
            if (gp[u].get(i).v == v)
                return i;
        }
        return -1;
    }

    public static void removeEdge(ArrayList<Edge>[] gp, int u, int v) {
        int temp = findEdge(gp, u, v);
        int temp_ = findEdge(gp, v, u);
        gp[u].remove(temp);
        gp[v].remove(temp_);
    }

    public static void removeVtx(ArrayList<Edge>[] gp, int vtx) {
        for (int i = gp[vtx].size() - 1; i >= 0; i--) {
            int temp = gp[vtx].get(i).v;
            removeEdge(gp, temp, vtx);
        }
    }

    public static void display(ArrayList<Edge>[] gp) {
        for (int i = 0; i < gp.length; i++) {
            System.out.print(i + "----> ");
            for (Edge e : gp[i])
                System.out.print("(" + e.v + " " + e.w + ") ,");

            System.out.println();
        }
    }

    public static void constructGraph() {
        int n = 7;
        ArrayList<Edge>[] graph = new ArrayList[n];
        for (int i = 0; i < graph.length; i++) {
            graph[i] = new ArrayList<>();
        }
        addEdge(graph, 0, 1, 10);
        addEdge(graph, 0, 3, 50);
        addEdge(graph, 1, 2, 10);
        addEdge(graph, 2, 3, 10);
        addEdge(graph, 3, 4, 10);
        addEdge(graph, 4, 5, 10);
        addEdge(graph, 4, 6, 10);
        addEdge(graph, 5, 6, 10);
        // addEdge(graph, 2, 5, 10);
        // removeEdge(graph, 3, 2);
        // removeVtx(graph, 4);
        // display(graph);
        boolean[] vis = new boolean[n];
        // System.out.print(allPath(0, 6, graph, vis,""));
        // hamiltonianPath(2, 2, graph, 0, vis, "");
        // System.out.println(nofCycles(graph, vis));
        // allPath(0,6, graph, vis, "");
        bfs(graph, 1, vis);

    }

    public static boolean hasPath(int src, int dest, ArrayList<Edge>[] gp, boolean[] vis) {
        if (src == dest) {
            return true;
        }
        vis[src] = true;
        boolean res = false;
        for (Edge e : gp[src])
            if (!vis[e.v])
                res = res || hasPath(e.v, dest, gp, vis);
        vis[src] = false;
        return res;
    }

    public static int allPath(int src, int dest, ArrayList<Edge>[] gp, boolean[] vis, String ans) {
        if (src == dest) {
            System.out.println(ans + dest + " ");
            return 1;
        }
        vis[src] = true;
        int count = 0;
        for (Edge e : gp[src])
            if (!vis[e.v])
                count = count + allPath(e.v, dest, gp, vis, ans + src + " ");
        vis[src] = false;
        return count;
    }

    // saari vertices ghum lo without repetion of any edge
    public static void hamiltonianPath(int src, int osrc, ArrayList<Edge>[] gp, int count, boolean[] vis, String ans) {
        if (count == gp.length - 1) {
            int t = findEdge(gp, src, osrc);
            if (t == -1) {
                System.out.println("Path" + ans + src);
            } else {
                System.out.println("Cycle" + ans + src);
            }
            return;
        }
        vis[src] = true;
        for (Edge e : gp[src]) {
            if (!vis[e.v])
                hamiltonianPath(e.v, osrc, gp, count + 1, vis, ans + src + " ");
        }
        vis[src] = false;
    }

    public static void dfs_GCC(ArrayList<Edge>[] gp, boolean[] vis, int src) {
        vis[src] = true;
        for (Edge e : gp[src]) {
            if (!vis[e.v])
                dfs_GCC(gp, vis, e.v);
        }
    }

    // this returns no of parts of a graph
    public static int nofCycles(ArrayList<Edge>[] gp, boolean[] vis) {
        int count = 0;
        for (int i = 0; i < gp.length; i++) {
            if (!vis[i]) {
                count++;
                dfs_GCC(gp, vis, i);
            }

        }
        return count;
    }

    public static class Pair {
        int vtx = 0;
        String ans = "";
        int level = 0;

        Pair(int vtx, String ans, int level) {
            this.vtx = vtx;
            this.ans = ans;
            this.level = level;
        }

        Pair(int vtx, String ans) {
            this.vtx = vtx;
            this.ans = ans;
        }

    }

    public static void bfs(ArrayList<Edge>[] gp, int src, boolean[] vis) {
        LinkedList<Pair> que = new LinkedList<>();
        int dest = 6;
        que.addLast(new Pair(src, src + " "));
        que.addLast(null);
        while (que.size() != 1) {
            if (que.getFirst() == null) {
                que.removeFirst();
                que.addLast(null);
            }
            Pair idxpair = que.removeFirst();
            if (vis[idxpair.vtx]) {
                System.out.println("Cycle: " + idxpair.vtx);
                continue;
            }
            if (idxpair.vtx == dest) {
                System.out.println("Dest: " + idxpair.ans);
            }
            vis[idxpair.vtx] = true;
            for (Edge e : gp[idxpair.vtx]) {
                if (!vis[e.v])
                    que.addLast(new Pair(e.v, idxpair.ans + e.v + " "));
            }

        }
    }

    public static void bfs_02(ArrayList<Edge>[] gp, int src, boolean[] vis) {
        LinkedList<Pair> que = new LinkedList<>();
        int dest = 6;
        que.addLast(new Pair(src, src + " ", 0));
        while (que.size() != 0) {
            Pair rvtx = que.removeFirst();
            if (vis[rvtx.vtx]) {
                System.out.println("Cycle: " + rvtx.ans);
                continue;
            }
            if (rvtx.vtx == dest) {
                System.out.println("Destination : " + rvtx.ans + " ---> " + rvtx.level);
            }
            vis[rvtx.vtx] = true;
            for (Edge e : gp[rvtx.vtx]) {
                if (!vis[e.v])
                    que.addLast(new Pair(e.v, rvtx.ans + e.v + " ", rvtx.level + 1));
            }
        }

    }

    public static void main(String[] args) {
        constructGraph();

    }
}

File: graphs/L00.cpp
#include <iostream>
#include <vector>
#include <queue>
using namespace std;

class Edge
{
public:
    int v = 0;
    int w = 0;
    Edge(int v, int w)
    {
        this->v = v;
        this->w = w;
    }
};
int n = 7;
vector<vector<Edge>> graph(n, vector<Edge>());

void addEdge(vector<vector<Edge>> gp, int u, int v, int w)
{
    graph[u].push_back(Edge(v, w));
    graph[v].push_back(Edge(u, w));
}

int findEdge(int v1, int v2)
{
    int i = 0;
    for (i = 0; i < graph[v1].size(); i++)
    {
        if (graph[v1][i].v == v2)
            break;
    }
    return i;
}

void removeEdge(int u, int v)
{
    int idx = findEdge(u, v);
    int idx_ = findEdge(v, u);
    graph[u].erase(graph[u].begin() + idx);
    graph[v].erase(graph[v].begin() + idx_);
}

void removeVtx(int vtx)
{
    for (int i = graph[vtx].size() - 1; i >= 0; i--)
    {
        removeEdge(vtx, graph[vtx][i].v);
    }
}

void display(vector<vector<Edge>> &gp)
{
    for (int i = 0; i < gp.size(); i++)
    {
        cout << i << "--> ";
        for (Edge e : gp[i])
        {
            cout << "(" << e.v << ", " << e.w << "), ";
        }
        cout << endl;
    }
}

// questions==================================================================================
bool hasPath(int src, int des, vector<bool> &vis)
{
    if (src == des)
    {
        return true;
    }
    vis[src] = true;
    bool res = false;
    for (Edge e : graph[src])
    {
        res = res || hasPath(e.v, des, vis);
    }
    vis[src] = false;
    return res;
}

int allPath(int src, int dest, vector<bool> &vis, int w, string ans)
{
    if (src == dest)
    {
        cout << ans << dest << " @ " << w << endl;
        return 1;
    }
    vis[src] = true;
    int count = 0;
    for (Edge e : graph[src])
    {
        if (!vis[e.v])
            count += allPath((e.v), dest, vis, e.w, ans + to_string(src) + " ");
    }
    vis[src] = false;
    return count;
}
void preOrder(int src, vector<bool> &vis, int w, string ans)
{
    vis[src] = true;
    cout << ans << " @ " << w << endl;
    for (Edge e : graph[src])
    {
        if (!vis[e.v])
            preOrder(e.v, vis, e.w, ans + to_string(e.v) + " ");
    }
    vis[src] = false;
}

void hamintonianPath(int src, int osrc, vector<bool> &vis, int count, string ans)
{
    if (count == vis.size() - 1)
    {
        for (Edge m : graph[src])
        {
            if (m.v == osrc)
            {
                cout << "Cycle  " << ans + to_string(src) << endl;
                return;
            }
        }
        cout << "Path  " << ans + to_string(src) << endl;
        return;
    }
    vis[src] = true;
    for (Edge e : graph[src])
    {
        if (!vis[e.v])
            hamintonianPath(e.v, osrc, vis, count + 1, ans + to_string(src) + " ");
    }
    vis[src] = false;
}

int GCC_dfs(int src, vector<bool> vis)
{
    int count = 0;
    vis[src] = true;
    for (Edge e : graph[src])
    {
        if (!vis[e.v])
            count += GCC_dfs(e.v, vis);
    }
    return count + 1;
}

int GCC()
{
    vector<bool> vis(n, false);
    int count = 0;
    int maxsize = 0;
    for (int i = 0; i < n; i++)
    {
        if (!vis[i])
        {
            count++;
            maxsize = max(maxsize, GCC_dfs(i, vis));
        }
    }
    cout << maxsize << endl;
    return count;
}

void bfs(int src, vector<bool> vis)
{
    queue<pair<int, string>> que;
    int dest = 6;
    que.push({src, to_string(src) + " "});
    while (que.size() != 0)
    {
        pair<int, string> temp = que.front();
        que.pop();
        if (vis[temp.first])
        {
            cout << "Cycle: " << temp.second << endl;
            continue;
        }
        if (temp.first == dest)
        {
            cout << "Destination: " << temp.second << endl;
        }
        vis[temp.first] = true;
        for (Edge e : graph[temp.first])
            if (!vis[e.v])
                que.push({e.v, temp.second + to_string(e.v) + " "});
    }
}
// push kr de rhe hai queue mein aur baad mein hm usko true kr rhe hai function mein jaane ke baad
// true mark kro aur uske childs ko push kro
void bfs_02(int src, vector<bool> vis)
{
    queue<int> que;
    int dest = 6;
    int level = 0;
    int cycle = 0;
    que.push(src);
    while (que.size() != 0)
    {
        int p = que.size();
        while (p-- > 0)
        {
            int temp = que.front();
            que.pop();
            if (vis[temp])
            {
                cout << "Cycle No: " << cycle << "--" << temp << endl;
                cycle++;
                continue;
            }
            if (temp == dest)
            {
                cout << "Destination: " << level << "--" << temp << endl;
            }
            vis[temp] = true;
            for (Edge e : graph[temp])
                if (!vis[e.v])
                    que.push(e.v);
        }
        level++;
    }
}

// isme hm push krte hi true mark kr rhe hai ;
// push krke true mark kr do
void bfs_03(int src, vector<bool> vis)
{
    queue<int> que;
    int dest = 6;
    int level = 0;
    que.push(src);
    vis[src] = true;
    while (que.size() != 0)
    {
        int size = que.size();
        while (size-- > 0)
        {
            int idx = que.front();
            que.pop();
            if (idx == dest)
            {
                cout << "Dest : " << level << endl;
            }
            for (Edge e : graph[idx])
            {
                if (!vis[e.v])
                {
                    que.push(e.v);
                    vis[e.v] = true;
                }
            }
        }
        level++;
    }
}
void constructGraph()
{
    // for(int i=0;i<N;i++){
    //     vector<Edge*> a;
    //     graph.push_back(a);
    // }
    addEdge(graph, 0, 1, 10);
    addEdge(graph, 0, 3, 10);
    addEdge(graph, 1, 2, 10);
    addEdge(graph, 2, 3, 40);
    addEdge(graph, 3, 4, 2);
    addEdge(graph, 4, 5, 2);
    addEdge(graph, 4, 6, 3);
    addEdge(graph, 5, 6, 8);
    // addEdge(graph, 2, 5, 40);
    // removeEdge(3, 4);
    // addEdge(graph, 2, 5, 2);

    // display(graph);
    // cout << endl;
}
int main()
{
    vector<bool> vis(n, false);
    constructGraph();
    bfs_03(1, vis);
    // preOrder(0, vis, 0, to_string(0) + " ");
    // hamintonianPath(2, 2, vis, 0, "");
    return 0;
}